freshmaker/lightblue.py

```diff
@@ -1831,7 +1831,8 @@ def get_fixed_published_image(self, name, version, image_group, rpm_nvrs, conten
             # If it's not on the same repositories or the regex matched something unexpected, then
             # skip it
             candidate_image_group = self.describe_image_group(image)
-            if candidate_image_group != image_group:
+            if candidate_image_group.split('[')[0] != image_group.split('[')[0] or \
+                    candidate_image_group.split('[')[1][:-1] not in image_group.split('[')[1][:-1]:
```

Comment:
Summary of our discussion in the chat: it seems like there's actually nothing that we can use to improve this. I'd still like to review this better, but I don't have the time today unfortunately. @qixiang @apaplaus does anyone of you feel like reviewing this or have the time to do it? Thank you.

Reply:
1. First, this `candidate_image_group.split('[')[1][:-1] not in image_group.split('[')[1][:-1]` is actually checking in the opposite way, `image_group` is the parent image freshmaker will use before the replacement, it could be an unreleased internal image, and its value could be something like `parent-image-2.3-['internal/repo-a']`,  `candidate_image_group` is the published image whose value could be something like `parent-image-2.3-['internal/repo-a', 'public/repo-a', 'public/repo-b']`, so actually in case the parent image is unreleased, `candidate_image_group.split('[')[1][:-1] not in image_group.split('[')[1][:-1]` is always `False`, and the unreleased parent image will never be replaced.
2. I'm not 100% sure about replacing the image when `"internal/repo-a"` is subset of `"internal/repo-a", "public/repo-a", "public/repo-b"`, though I don't see any exceptions based on what I have seen with these repos, so I'm fine with this change.
3. I agree this line is very error prone and hard for readability, maybe we can add a inline function to just extract the repos as a set from the described image group value, and add example comment to explain the purpose.

Reply:
1. ahh gotcha, the `candidate_image_group` and `image_group` are now swapped.
2. :+1:
3. I adjusted the code to make the repos into an iterable list (preventing an error from string parsing). The code becomes a little bit more difficult to read but I added a comment to describe what the code does.
(details here about the code: after checking the image group, the code extracts the repos into a list and iterates through each item in the image_group repo list and makes sure they all exist within candidate_image_group repo list)

Reply:
How about this:
```
def _extract_name_version_and_repos(image_group_string):
    # Extract name-version and repos from described string of image_group which
    # is generated by describe_image_group.
    # This helper function returns a tuple of (name-version-string, set-of-repos)
    # Example:
    # image_group_string:
    #   "parent-image-2.3-['internal/repo-a', 'public/repo-a', 'public/repo-b']"
    # returns:
    #   ('parent-image-2.3', {'internal/repo-a', 'public/repo-a', 'public/repo-b'})
    image_group_string = image_group_string.replace("'", "").replace('"', '')
    name_version, repos = re.search(r"(.*)-\[([^]]+)\]", image_group_string).groups()
    return (name_version, {x.strip() for x in repos.split(",")})

name_version, repos = _extract_name_version_and_repos(image_group)
candidate_name_version, candidate_repos = _extract_name_version_and_repos(candidate_image_group)

if not (name_version == candidate_name_version and repos.issubset(candidate_repos)):
    continue
```

Reply:
Beside of the previous suggestion, I think we can also change the `describe_image_group` to return a result that's easier for use later.

Reply:
@gnaponie @apaplaus wdyt? the current code or even with the helper function is error prone and not good for readability. How about change the `describe_image_group` to return something more easier to use?

It can be more structured string or even a class like this (maybe too heavy?):
```
class ImageGroup:
    def __init__(self, image, lightblue):
        parsed_nvr = koji.parse_NVR(image.nvr)
        repositories = image.get_registry_repositories(lightblue)
        self.name = parsed_nvr["name"]
        self.version = parsed_nvr("version")
        self.repos = {x["repository"] for x in repositories}

    def __eq__(self, other):
        return all(
            [
                self.name == other.name
                and self.version == other.version
                and self.repos == other.repos
            ]
        )

    def issubset(self, other):
        return all(
            [
                self.name == other.name
                and self.version == other.version
                and self.repos.issubset(other.repos)
            ]
        )


class Lightblue:
    ...

    def describe_image_group(self, image):
        return ImageGroup(image, self)
```

Reply:
I would vote for having a separate small helper class for this purpose. As I see it really becomes complicated and we would like to have it more reliable and readable. I think it's better if we have more lines of code with a new class, then just leave it in a few unreadable lines with 10 lines comment.

Reply:
so I tried it using the helper class to get the name version repo in a format that's less error-prone and more easily readable:
`{'name': 'rhel-server-container', 'version': '7.9', 'repos': {'repo'}}`
as opposed to the one we are currently using
`rhel-server-container-7.9-['repo']`

however, a ton of code in lightblue.py expects `describe_image_group` to return the name version repo in the string format that we are using --> `rhel-server-container-7.9-['repo']` 
If I'm not mistaken, changing it would require change to all the code using the describe_image_group function to cater to a dictionary vs a string.

Here are unit test screenshots to show which files are affected. The first screenshot is the original code and the second shows the drawback of changing `describe_image_group`

![Screenshot from 2021-11-05 01-03-11](https://user-images.githubusercontent.com/44615679/140461210-5f5d971e-7e02-4d48-b6bc-a828282cac5f.png)

![Screenshot from 2021-11-05 01-05-30](https://user-images.githubusercontent.com/44615679/140461216-0dc2a0ad-ec30-4557-b9de-6f2aba53dee0.png)

Reply:
Creating a helper class to get the `candidate_image_group` in the form we want makes sense and I'm able to get it in a dictionary like this --> `{'name': 'rhel-server-container', 'version': '7.9', 'repos': {'other_repo'}}`
instead of a string like this -->`rhel-server-container-7.9-['repo'])`

However, even if the `candidate_image_group` is in a dictionary format (to make it less error-prone), the `image_group` that we're comparing it to still comes as a string so we would have to use a string parser/regex anyway. Thoughts on this?

Reply:
Yes, with the helper class, you'll need to update some of the code accordingly, I think there is no too much to be changed.

> the image_group that we're comparing it to still comes as a string so we would have to use a string parser/regex anyway.
You don't need to compare with strings, the `ImageGroup` already has `__eq__` implemented. In case you need a string representation (for example use it as a dict key) you can implement `__str__` which returns the same string as before, and use `str(image_group)` to keep it compatible.

I don't have strong opinion of using the helper class, I'll leave it to you and other reviewers to decide.

Reply:
makes sense! changes have been made :+1: